d3/query-graphs.js

```diff
@@ -770,19 +815,89 @@ function drawQueryTree(target, treeData) {
             .duration(duration)
             .style("fill-opacity", 1)
             .attr("x", function(d) {
-                return (d.source.x + d.target.x) / 2;
+                return (d.source.x0 + d.target.x0) / 2;
             })
             .attr("y", function(d) {
-                return (d.source.y + d.target.y) / 2;
+                return (d.source.y0 + d.target.y0) / 2;
             });
 
         // Remove labels
         linkLabel.exit().transition()
             .duration(duration)
-            .attr("x", source.x)
-            .attr("y", source.y)
+            .attr("x", source.x0)
+            .attr("y", source.y0)
             .style("fill-opacity", 0)
             .remove();
+
+        // Update crosslinks
+        if (crosslinks !== undefined && crosslinks.length) {
+            var visibleCrosslinks = crosslinks.filter(function(d) {
+                return nodes.includes(d.source) && nodes.includes(d.target);
+            });
+            var crossLink = svgGroup.selectAll("path.crosslink")
+                .data(visibleCrosslinks);
+            crossLink.enter().insert("path", "g")
+                .attr("class", "crosslink")
+                .attr("d", function(_d) {
+                    var o = {
+                        x: source.x0,
+                        y: source.y0
+                    };
+                    return diagonalRaw({
+                        source: o,
+                        target: o
+                    });
+                });
+            crossLink.transition()
+                .duration(duration)
+                .attr("d", diagonal);
+            crossLink.exit().transition()
+                .duration(duration)
+                .attr("d", function(_d) {
+                    var o = {
+                        x: source.x,
+                        y: source.y
+                    };
+                    return diagonal({
+                        source: o,
+                        target: o
+                    });
+                })
+                .remove();
+
+            // Update highlighted crosslinks
```

Comment:
The code below seems to be duplicated with the code starting at line 839. The only difference I saw is the class name `crosslink` vs. `crosslink-highlighted`.
I think we should factor this into a helper function `updateCrosslinkPath(cssClass)` and then call it twice, once as `updateCrosslinkPath("crosslink")` and once as `updateCrosslinkPath("crosslink-highlighted")`

Reply:
Yes. Also needs opacity.